massa-protocol-worker/src/protocol_network.rs

```diff
@@ -0,0 +1,426 @@
+                        loss of sync between us and the remote node",
+                        node,
+                    );
+                    let _ = self.ban_node(&node).await;
+                }
+            }
+            NetworkEvent::ReceivedOperationAnnouncements {
+                node,
+                operation_prefix_ids,
+            } => {
+                massa_trace!(OPS_BATCH, { "node": node, "operation_ids": operation_prefix_ids});
+                self.on_operations_announcements_received(operation_prefix_ids, node)
+                    .await?;
+            }
+            NetworkEvent::ReceiveAskForOperations {
+                node,
+                operation_prefix_ids,
+            } => {
+                massa_trace!(ASKED_OPS, { "node": node, "operation_ids": operation_prefix_ids});
+                self.on_asked_operations_received(node, operation_prefix_ids)
+                    .await?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Network ask the local node for blocks
+    ///
+    /// React on another node asking for blocks informations. We can forward the operation ids if
+    /// the foreign node asked for `AskForBlocksInfo::Info` or the full operations if he asked for
+    /// the missing operations in his storage with `AskForBlocksInfo::Operations`
+    ///
+    /// Forward the reply to the network.
+    async fn on_asked_for_blocks_received(
+        &mut self,
+        from_node_id: NodeId,
+        list: Vec<(BlockId, AskForBlocksInfo)>,
+    ) -> Result<(), ProtocolError> {
+        let node_info = match self.active_nodes.get_mut(&from_node_id) {
+            Some(node_info) => node_info,
+            _ => return Ok(()),
+        };
+        let mut all_blocks_info = vec![];
+        for (hash, info_wanted) in &list {
+            let operations_ids = match self.storage.retrieve_block(hash) {
+                Some(wrapped_block) => wrapped_block.read().content.operations.clone(),
+                None => {
+                    // let the node know we don't have the block.
+                    all_blocks_info.push((*hash, ReplyForBlocksInfo::NotFound));
+                    continue;
+                }
+            };
+            let block_info = match info_wanted {
+                AskForBlocksInfo::Info => ReplyForBlocksInfo::Info(operations_ids),
+                AskForBlocksInfo::Operations(op_ids) => {
+                    // Mark the node as having the block.
+                    node_info.insert_known_blocks(
+                        &[*hash],
+                        true,
+                        Instant::now(),
+                        self.config.max_node_known_blocks_size,
+                    );
+
+                    // Send only the missing operations.
+                    let needed_ops = operations_ids
+                        .into_iter()
+                        .filter(|id| op_ids.contains(id))
```

Comment:
* no need to perform this `filter`
* first make sure that when the `AskForBlocksInfo::Operations` message is deserialized, we fail if there are too many ops being asked (constant config parameter)
* before we do `send_block_info` on response to `AskForBlocksInfo::Operations` we should first gather all the asked operations from storage, and if any is absent, send NotFound
* if all are found, send all the WrappedOperations throught send_block_info (and not just their IDs)

Reply:
@gterzian

Reply:
discussed in https://github.com/massalabs/massa/issues/2884